src/librustc_mir/shim.rs

```diff
@@ -259,6 +272,164 @@ impl<'a, 'tcx> DropElaborator<'a, 'tcx> for DropShimElaborator<'a, 'tcx> {
     }
 }
 
+/// Build a `Clone::clone` shim for `recvr_ty`. Here, `def_id` is `Clone::clone`.
+fn build_clone_shim<'a, 'tcx>(tcx: ty::TyCtxt<'a, 'tcx, 'tcx>,
+                              def_id: DefId,
+                              rcvr_ty: ty::Ty<'tcx>)
+                              -> Mir<'tcx>
+{
+    let sig = tcx.fn_sig(def_id);
```

Comment:
Small note: this is a "fully generic" shim, so it knows the concrete receiver type. You could check if the source type is Copy and emit an "optimized" impl that just copies the result. This will speed up cloning `[u8; 1048576]` on unoptimized builds and is generally a good idea.

Also, as @eddyb says, it might be a smart idea to plumb `#[derive(Clone)]` impls through this to avoid implementing the logic twice and to potentially speed up compilation - maybe have `#[derive(Clone)]` add a `#[is_builtin_clone]` attribute to the method?